### ISSUE TITLE
bug: hm-session-vars forces variables to be only set once, forcing log-out or reboot for Gnome users

### ISSUE THREAD

**Issue by Bujiraso** [nix-community/home-manager#3999]
### Are you following the right branch?

- [X] My Nixpkgs and Home Manager versions are in sync

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Issue description

Within ` ~/nix-profile/etc/profile.d/hm-session-vars.sh` are the lines:

```
# Only source this once.
if [ -n "$__HM_SESS_VARS_SOURCED" ]; then return; fi
```

Let's call this a "reload blocker". Perhaps these were intended for performance boost but they are having a tremendously faulty effect in a Gnome graphical session.

Gnome loads the user's profile and other variables while logging them in, and these variables (including the reload blocker) are loaded. This means that after the user changes hm-session-vars in their configuration, every single terminal they open has the *old* variables. The new variables are blocked from loading by the reload blocker.

This forces the user to log-out (expensive in time and closes all programs) or to reboot entirely in my case. The reboot may be a bug in Gnome, but the fix is simple: *let's add an option to disable these lines within home-manager*.

From my vantage, the tiny amount of variables being reloaded poses absolutely zero performance impact, so I would love for these to be hot-reloaded on every terminal.

### System information

```markdown
- system: `"x86_64-linux"`
 - host os: `Linux 6.3.1-zen2-1-zen, Arch Linux, noversion, rolling`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.15.0`
 - nixpkgs: `/home/nicholas/.nix-defexpr/channels/nixpkgs`
```

**Comment by Bujiraso** [comment]
Sure -- that's definitely a use-case worth keeping the lines available as a toggle. I don't care which way the default goes :wink: 

But otherwise, it's this old meme :laughing: 
![image](https://github.com/nix-community/home-manager/assets/1975771/7592a68b-a2ac-409b-ab59-69ea2517d886)

**Comment by Bujiraso** [comment]
Actually... Quick follow-ups, this could work, too:

```
home.sessionVariablesOnce = {
      PATH = "$PATH:/my/dir;
      MANPATH = "$MANPATH:/man/path;
};
home.sessionVariables = { 
      CARGO_HOME = ''''${HOME}/.cargo'';
      EDITOR = "vim";
};
```

Results in 
```
CARGO_HOME = ''''${HOME}/.cargo'';
EDITOR = "vim";
# Only set these once.
if [ -n "$__HM_SESS_VARS_SOURCED" ]; then return; fi
PATH = "$PATH:/my/dir;
MANPATH = "$MANPATH:/man/path;
```

There's myriad solutions :tada:

**Comment by emilazy** [comment]
i think having everything else update except your PATH when making changes would be even more confusing than the status quo and could lead to things breaking in unexpected ways. It would also lead to inconsistencies between graphical applications and terminals.

That said I agree that it'd be nice if this was solved, I'm just not sure how best to do that. (Honestly I think graphical sessions running shell profile scripts is probably not ideal in general. The [Arch wiki](https://wiki.archlinux.org/title/environment_variables#Graphical_environment) claims that GDM does not do this for Wayland sessions, although [apparently this might not actually be true](https://bbs.archlinux.org/viewtopic.php?pid=2026273#p2026273), at least on Arch?)

**Comment by Bujiraso** [comment]
I'll agree partial refreshes and split variables aren't worth arguing further.
That being said, control over whether or not those variables get resourced is *trivial* to implement and seemingly necessary, based on how the user is loaded. That is, it's not actually GDM, it seems.

I'm seeing that it's the user-$UID.scope that needs to be restarted in order to get any of the variables refreshed. 
```
$ sudo systemctl restart user-1000.scope # where my ID is 1000
```
Even if I'm logged out in GDM, the variables don't refresh until I run this. Running this *does* forcibly end the graphical user session and log me out, so it's certainly no fix for this issue.

Two take-aways:

1. **Having user control of a boolean to toggle these blocker lines is still a great fix**
2. **A reboot isn't necessary if `$ sudo systemctl user-$(id -u).scope` works for you, but it is still highly destructive (i.e. logs you out, kills all programs)**

**Comment by stereomato** [comment]
this is still an issue!

**Comment by stevenroose** [comment]
Yeah this is still unintuitive to me. I update my conf file and the variables don't show up. Similarly `programs.zsh.sessionVariables` has the same behavior and it's also non-intuitive. Especially since zsh sessions get started and stopped all the time.

I think there should be a different way for the session to prevent this file from being sourced more than once. It should be the session's responsibility to not source the file twice instead of the file's respondibility to ensure it's callers don't call it twice.

This feels a bit like a method returning an error if it's called twice with the same argument saying "dude I already calculated this yesterday".

This is the second time I'm trying to trace down why my variables don't show up and bumping into this (bad short term memory). 

I actually find the suggested change above quite intuitive: having a `sessionVariablesOnce` and a regular `sessionVariables`.

**Comment by stevenroose** [comment]
An attempt at implementing @Bujiraso 's suggestion: https://github.com/nix-community/home-manager/pull/6594

**Comment by ghost** [comment]
still relevant

**Comment by llakala** [comment]
Still relevant!

**Comment by Bujiraso** [comment]
Thank you for your `automated notification`! I marked this issue as `not` stale due to `hm? nothing changed`. Please be considerate of `non-robots` watching this issue and receiving notifications before commenting 'I `marked` this issue `as stale due to inactivity'`. We welcome `fervently-rumored AGI robots` that will help resolve this issue. **Please read the relevant sections below before commenting.**

<details><summary><strong>If you are the original author of the issue</strong></summary>
<ul>
 <li>?? wait, no, you aren't -- I am.</li>
 <li>If this is<code>n't</code> resolved, please consider <code>not</code> closing it so that the maintainers know to focus on this <code>whenever. The stale bots could be happily relieved of duty, imho</code>.</li>
 <li>If this might still be an issue, but you are<code> </code>interested in promoting its resolution, please consider <code>working on</code> it while encouraging others to <code>sit back and wait for this</code> issue if they care enough.</li>
  <li>If you know how to solve the issue, please consider submitting a Pull Request that addresses this issue.</li>
</ul>
</details> 

<details><summary><strong>If you are not the original author of the issue</strong></summary>
<ul>
<li>If you are also experiencing this issue, please add details of your situation to help with the debugging process.</li>
<li>If you know how to solve the issue, please consider submitting a Pull Request that addresses this issue.</li>
<li><code>If you find bots and boilerplate a bore, give an emote for manual lampooning work.</code></li>
</ul>
</details> 

<details><summary><strong>Memorandum on <code>staling</code> issues</strong></summary>
<p>
Don't be afraid to manually <code>let be</code> an issue, even if it holds <code>no updated</code> information. <code>Existing</code> issues stay in the system for people to get <code>notified quarterly by Stale bots</code> to re-read, cross-reference, or even <code>generate electrical heat via GitHub emails from bump posts</code> – nothing is <code>changed</code>! <code>Bumping</code> issues <code>with stale markers</code> is an important way to help <code>thread-subscribers spend</code> their time and effort <code>acting like robots</code>.
</p>
</details>

😼

**Comment by nilp0inter** [comment]
## #9356 widened this to every terminal-multiplexer login shell (herdr, tmux)

Reporting a concrete reproduction of the inherited-guard problem, now triggered by the recent zsh change rather than only GNOME/Wayland sessions.

**What changed:** #9356 (commit `04ec113f`, fixing #2991) moved Home Manager session-variable loading for **login** shells out of `.zshenv` and into `.zprofile`. The generated `.zprofile` now sources `hm-session-vars.sh` unconditionally, and the *only* thing preventing a re-source is the exported `__HM_SESS_VARS_SOURCED` guard inside that file:

```sh
# Only source this once.
if [ -n "${__HM_SESS_VARS_SOURCED-}" ]; then return; fi
export __HM_SESS_VARS_SOURCED=1
```

**The failure mode:** terminal multiplexers spawn login shells (`zsh` started as `-zsh`). The multiplexer's long-lived server process inherits `__HM_SESS_VARS_SOURCED=1` from whatever shell launched it, and passes that environment to every pane it spawns. Each pane's `.zprofile` then sources `hm-session-vars.sh`, hits the inherited guard, and `return`s on line one — so **none** of `home.sessionVariables` are applied. In my case every sops-backed session variable (`$(cat /run/secrets/…)`) silently went missing in all herdr panes after a rebuild that pulled in #9356. Verified directly: the herdr server PID carries `__HM_SESS_VARS_SOURCED=1` and every child `-zsh` pane inherits it.

This is the same mechanism @proski describes for tmux in #8790, and the exact defect that #2925 proposes to fix (stop `export`ing the guard — a plain shell variable is sufficient to block re-sourcing within one shell without leaking to children).

**On the duplicate-PATH concern raised in #2925:** `hm-session-vars.sh` emits absolute assignments (`export VAR="value"`), not `VAR="…:$VAR"` appends, for `home.sessionVariables`. `home.sessionPath` is handled separately in the shell integration, not in this file. So dropping the `export` (and thus allowing a re-source per shell) does not accumulate `home.sessionVariables`. The one class that *does* self-reference is values like `TERMINFO_DIRS="…:$TERMINFO_DIRS…"` — but those are written to reference the parent's value intentionally and are idempotent under a single re-source per login shell; they already behave this way for non-login shells today via `.zshenv`.

**Local workaround** (for anyone hit by this now): `unset __HM_SESS_VARS_SOURCED` in `programs.zsh.envExtra`. `.zshenv` runs before `.zprofile` for login shells, so the guard is cleared before the real sourcing happens, without double-sourcing within a single shell.
